tests/unit/s2n_send_test.c

```diff
@@ -37,37 +45,122 @@ static int s2n_track_sent_bytes_fn(void *io_context, const uint8_t *buf, uint32_
+}
+
+/* Mock send that verifies all sends are the same size as the max amount of bytes that fit in a single Ethernet frame. */
+static int s2n_dynamic_record_sizing_fn(void *io_context, const uint8_t *buf, uint32_t len)
+{
+    struct s2n_connection *conn = (struct s2n_connection*) io_context;
+    s2n_custom_send_fn_called = true;
+    uint32_t *writes = (uint32_t*) s2n_connection_get_ctx(conn);
+    
+    uint16_t single_eth_frame_record_size = 0;
+    POSIX_GUARD_RESULT(s2n_record_min_write_payload_size(conn, &single_eth_frame_record_size));
+
+    /* Until we hit the dynamic record resize threshold we expect that the total data sent in bytes is divisible
+     * by the s2n_record_min_write_payload_size. Once the threshold is exceeded the feature is no longer active 
+     * until the dynamic record resize timer is triggered.
+     * 
+     * Since the send is larger than the dynamic record resize threshold and the connection timer for the dynamic sizing
+     * is in practice frozen for this test we can control which writes using the dynamic record sizing. 
+     *
+     * We will expect that the first write and the third write are forced to use dynamic record sizing, so every
+     * record fragment sent over the wire should be divisble by single_eth_frame_record_size. We check this property 
+     * by making sure conn->current_user_data_consumed modulus single_eth_frame_record_size is equal to zero. */
+    if (conn->active_application_bytes_consumed <= conn->dynamic_record_resize_threshold) {
+        /* The test is set up so dynamic record sizes are only used in the first and third writes. */
+        EXPECT_TRUE(*writes == 1 || *writes == 3);
+        EXPECT_EQUAL(conn->current_user_data_consumed % single_eth_frame_record_size, 0);
+    }
+
+    /* We don't care about the send sizes once we have exceeded conn->dynamic_record_resize_threshold. */
+    EXPECT_SUCCESS(s2n_stuffer_wipe(&conn->out));
+    EXPECT_SUCCESS(s2n_stuffer_write_bytes(&conn->out, buf, len));
+
+    return len;
+}
+
+/* Mock send implementation that will return EAGAIN once a specified limit is hit. */
+static int s2n_byte_limit_send_fn(void *io_context, const uint8_t *buf, uint32_t len)
+{
+    struct s2n_connection *conn = (struct s2n_connection*) io_context;
+    s2n_custom_send_fn_called = true;
+    ssize_t *bytes_to_send = (ssize_t*) s2n_connection_get_ctx(conn);
+
+    /* Cede control back to application when the bytes_to_send threshold is reached. */
+    if (conn->current_user_data_consumed > *bytes_to_send) {
+        errno = EAGAIN;
+        return -1;
+    }
+
+    EXPECT_SUCCESS(s2n_stuffer_wipe(&conn->out));
+    EXPECT_SUCCESS(s2n_stuffer_write_bytes(&conn->out, buf, len));
+
+    return len;
+}
 
 int main(int argc, char **argv)
 {
     BEGIN_TEST();
 
+    DEFER_CLEANUP(struct s2n_blob large_data = {0}, s2n_free);
+    /* Allocate 32KB so it takes multiple records to send complete data. */
+    EXPECT_SUCCESS(s2n_alloc(&large_data, 1 << 15));
```

Comment:
Don't we have a constant somewhere representing the maximum record size? Could we use that here rather than "<<15"?

Reply:
This integer is larger than the maximum record size so the message is chunked across multiple records.

I'll move this into a macro